[IndexedDB] An array keyPath should yield an array key

cdumez opened the pull request:
#### 54a7b93842dd440a06ca58f89e15c4206c129772
<pre>
[IndexedDB] An array keyPath should yield an array key
<a href="https://bugs.webkit.org/show_bug.cgi?id=257238">https://bugs.webkit.org/show_bug.cgi?id=257238</a>

Reviewed by Sihui Liu.

An array keyPath should yield an array key, to match the behavior of Blink and Gecko.

* LayoutTests/imported/w3c/web-platform-tests/IndexedDB/idbindex_keyPath.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/IndexedDB/idbindex_keyPath.any.js:
* LayoutTests/imported/w3c/web-platform-tests/IndexedDB/idbindex_keyPath.any.worker-expected.txt:
Resync WPT test from upstream to get test coverage. Without this change, the
second subtest was failing with WebKit (but passing with Blink &amp; Gecko).

* Source/WebCore/Modules/indexeddb/shared/IndexKey.cpp:
(WebCore::IndexKey::IndexKey):
(WebCore::IndexKey::asOneKey const):
(WebCore::IndexKey::multiEntry const):
* Source/WebCore/Modules/indexeddb/shared/IndexKey.h:
(WebCore::IndexKey::isNull const):
* Source/WebCore/bindings/js/IDBBindingUtilities.cpp:
(WebCore::createKeyPathArray):
(WebCore::generateIndexKeyForValue):
IndexKey was losing was storing the key as a Vector unconditionally and thus
was losing the information about whether or not the key was a vector or not.
To address the issue, we now use a `std::variant&lt;std::nullptr_t, IDBKeyData, Vector&lt;IDBKeyData&gt;&gt;`
with nullptr_t being the null state that this class supports.

Canonical link: <a href="https://commits.webkit.org/264479@main">https://commits.webkit.org/264479@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/983c475d58362ffc2541e273e75fb8f01661d2aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7751 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8023 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8203 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9393 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7906 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7757 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10013 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7944 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/10780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7886 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9001 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7116 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9503 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6309 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7055 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7460 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7181 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10595 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7675 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6255 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7002 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1854 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11214 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7406 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/Modules/indexeddb/shared/IndexKey.h:
```diff
@@ -34,18 +34,18 @@ namespace WebCore {
 
 private:
-    Vector<IDBKeyData> m_keys;
+    std::variant<std::nullptr_t, IDBKeyData, Vector<IDBKeyData>> m_keys;
```
Can we add type alias for this? e.g. `using Data = std::variant<std::nullptr_t, IDBKeyData, Vector<IDBKeyData>>;`

cdumez replied:
Good idea, will do

szewai reviewed (approved):
r=me

cdumez committed:
[IndexedDB] An array keyPath should yield an array key
https://bugs.webkit.org/show_bug.cgi?id=257238

Reviewed by Sihui Liu.

An array keyPath should yield an array key, to match the behavior of Blink and Gecko.

* LayoutTests/imported/w3c/web-platform-tests/IndexedDB/idbindex_keyPath.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/IndexedDB/idbindex_keyPath.any.js:
* LayoutTests/imported/w3c/web-platform-tests/IndexedDB/idbindex_keyPath.any.worker-expected.txt:
Resync WPT test from upstream to get test coverage. Without this change, the
second subtest was failing with WebKit (but passing with Blink & Gecko).

* Source/WebCore/Modules/indexeddb/shared/IndexKey.cpp:
(WebCore::IndexKey::IndexKey):
(WebCore::IndexKey::asOneKey const):
(WebCore::IndexKey::multiEntry const):
* Source/WebCore/Modules/indexeddb/shared/IndexKey.h:
(WebCore::IndexKey::isNull const):
* Source/WebCore/bindings/js/IDBBindingUtilities.cpp:
(WebCore::createKeyPathArray):
(WebCore::generateIndexKeyForValue):
IndexKey was losing was storing the key as a Vector unconditionally and thus
was losing the information about whether or not the key was a vector or not.
To address the issue, we now use a `std::variant<std::nullptr_t, IDBKeyData, Vector<IDBKeyData>>`
with nullptr_t being the null state that this class supports.

Canonical link: https://commits.webkit.org/264479@main
